### ISSUE TITLE
Testing for scale

### ISSUE THREAD

**Issue by mikefab** [unicef/magicbox-multihop-mobility#1]
@computate, @d-kaylor, @bryantson, thank you so much for participating in the [challenge](https://github.com/unicef/magicbox-kepler-demo/issues/18) yesterday.

@marcellamaki took at crack at this last week, [here]( https://gist.github.com/marcellamaki/93842d718feda2680a24c169dc35f2fb) is her solution. It iterates through each user, calculating and inserting their multi-hop trips in a table, before aggregating all trips.

We're now testing whether these solutions can handle millions of rows.

@vedransekara has prepared [a call test dataset](https://github.com/unicef/magicbox-multihop-mobility/blob/master/sample/dummy_call_data.csv.zip), consisting of call data for 1000 people for one year and 26 different cell towers, 833k event in total.

Are there any edits or other strategies you might recommend for running this type of query on ORACLE 11 or SQL Server 2008?

**Comment by computate** [comment]
Right, I was pondering that myself. The first thing I would do is process each sim_id individually if performance is terrible, which it might. So start by adding this to the very inner query ( HAVING m1.sim_id=?) and you can obtain the sim_id list pretty easily: select distinct site_id from mobility; 

```sql
SELECT site1, site2, count(*) num FROM (
	SELECT sim_id, site1, site2 FROM (
		SELECT m3.sim_id, m4.site_id site1, m5.site_id site2, m3.created1, m3.created2 FROM (
				SELECT m1.sim_id, m1.created created1, m2.created created2 FROM mobility m1 
				INNER JOIN mobility m2 on m1.sim_id = m2.sim_id AND m1.created < m2.created 
					GROUP BY m1.sim_id, m1.created, m2.created HAVING m1.sim_id=? ORDER BY m1.sim_id, m1.created
			) m3 
			INNER JOIN mobility m4 ON m3.sim_id = m4.sim_id AND m3.created1 = m4.created 
			INNER JOIN mobility m5 ON m3.sim_id = m5.sim_id AND m3.created2 = m5.created 
			WHERE m4.site_id != m5.site_id
			ORDER BY m3.sim_id, m3.created1, m3.created2
		) m6 
		GROUP BY m6.sim_id, m6.site1, m6.site2, m6.site1, m6.site2 HAVING m6.site1 != m6.site2
	) m7 GROUP BY m7.site1, m7.site2
;
```

If there is still more performance problems, you may consider creating a temporary table for the inner query: 
```sql
insert into mobility_hops select m1.sim_id, m1.created created1, m2.created created2 FROM mobility m1 
				INNER JOIN mobility m2 on m1.sim_id = m2.sim_id AND m1.created < m2.created 
					GROUP BY m1.sim_id, m1.created, m2.created;
```

You can basically improve efficiency from the inside of the query out, just like you would most easily understand the query from the inside out. But keep me posted if there are additional challenges to work through.

**Comment by marcellamaki** [comment]
Hi @computate. The initial solution I worked on ([here](https://gist.github.com/marcellamaki/93842d718feda2680a24c169dc35f2fb)), uses temporary tables. I was wondering if you might take a look at see if you think it's in line with what you are suggesting, and if you have any additional ideas for this type of optimization?
